airflow/dag_processing/processor.py

```diff
@@ -837,13 +877,17 @@ def process_file(
         :return: number of dags found, count of import errors, last number of db queries
         """
         self.log.info("Processing file %s for tasks to queue", file_path)
+        try:
+            if InternalApiConfig.get_use_internal_api():
+                dagbag = DagFileProcessor._get_dagbag(file_path)
+            else:
+                with create_session() as session:
```

Comment:
I am curious. You create a session here but you seem not using it (besides for the listener)?

Reply:
Yeah. I have to take a closer look at that one - I think I made wrong assumption here about where the queries were counted and I even commented it out - but I see now it's quite a bit differrent counting

Reply:
All right @jscheffl and @vincbeck . I think I have it right - I will do some more manual testing and see if it works in various combinations. Looking for your comments.

But the things I found that the counter was LIKELY not doing what it was supposed to do when added in https://github.com/apache/airflow/pull/40323 

@MaksYermak @kosteev @VladaZakharova -> this is probably last "serious" change to make AIP-44 work (still special tests and docs will be needed but this can be added after that). Let me know if I understood it properly.

When I looked at the counter from #40323 the event was added to the session just before creating dagbag from the file - apparently with the intention to cound the DB queries done inside the dagbag parsing. But as far as I uniderstend (please correct me if I am wrong) ... It would not work because the session the event has been registered for is not passed down/used in the dagbag parsing - any DB access there would create a new session rather than use the session created in "process_file" - so it would not count any queries there. But it was indeed counting potentially queries run during callback, updating warning etc.

I **THINK** that was not the intention of that change - the intention was to count queries resuting from parsing the DAG (but it will not). I am not sure if we can easily fix it (but it does not matter in DB isolation mode at least because no DB queries will be allowed when parsing the DAG anyway). 

In this change I attempted to retain that behaviour. Due to the fact that we cannot pass parsed dag to Internal-api component, the whole "process_file" is now effectively split into more transactons - 4 instead of 2 as it was originally (but I think that is ok, transactional integrity is not important in those cases - and I collect the count of queries and add them up from all 4 calls that might turn into "internal_api_call" - so DB queries will be counted in remote end and passed back as return value of those calls. So the metrics should remain the same as it was.

I introduced a context manager for that - that is nicer than the original method decorated. Speaking of which I also found one potential memory leak. I think when you register for the events, you should always remove the event function because otherwise it might leak memory (see the docstring). It would not matter much in this case - because parsing file is always done in a fork, but I did it in the way that context manager cleans up and removes such registered event - in "database isolation" case it would matter because there memory leaks could stay in the gunicorn threads handling the calls. 

Let me know if the lines of thought above make sense please :)

Reply:
Makes sense to me. Thanks for sharing the details. I like the context manager solution

Reply:
@MaksYermak @kosteev @VladaZakharova -> before I merge this one, please let me know if I correctly assessed the event behaviour. Was the intention to count the queries made by the DAG parsing as well ? Because if it was, then I am afraid it's not going to work. The only possibility for that would be to make some other event (not sure if possible) that will attach such event handler to any new session created during DAG parsing. Not crucial for this PR to get merged, but would be great to see if this is actually ok according to intentions.

Reply:
@potiuk thank for your investigation! For me your message makes sense. You are right the original idea was to count all queries during DAG file processing. And maybe now I understand why with different event handlers the number of queries are different. I have mentioned it here: https://github.com/apache/airflow/pull/40323#discussion_r1650979649

I have made investigations and recorded queries for event handlers with `session` object and `engine` object. Here's the result:
```
Recorded query locations session:
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:158:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:168:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:181:	5
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1807 > override.py:get_action:1676:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1808 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1745 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1840 > override.py:create_action:1685 > override.py:get_action:1676:	15
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1811:	10
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3171:	1
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3191:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3194 > dagrun.py:active_runs_of_dags:380:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3263 > dagcode.py:bulk_sync_to_db:82:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:618:	2
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:625:	1
	test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:689 > processor.py:_validate_task_pools:668 > pool.py:get_pools:69:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:691:	1

Recorded query locations engine:
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:158:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:168:	5
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:672 > serialized_dag.py:write_dag:181:	5
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1807 > override.py:get_action:1676:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1808 > override.py:get_resource:1737:	15
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1745 > override.py:get_resource:1737:	15
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1839 > override.py:create_resource:1751:	10 -> new
	dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1840 > override.py:create_action:1685 > override.py:get_action:1676:	15
	dagbag.py:_sync_to_db:707 > dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1845:	15 -> new
	dagbag.py:_serialize_dag_capturing_errors:679 > dagbag.py:_sync_perm_for_dag:735 > override.py:sync_perm_for_dag:1101 > override.py:create_permission:1836 > override.py:get_permission:1811:	10
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3171:	1
	test_processor.py:_process_file:114 > processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3191:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3194 > dagrun.py:active_runs_of_dags:380:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3263 > dagcode.py:bulk_sync_to_db:82:	1
	processor.py:process_file:883 > processor.py:save_dag_to_db:919 > dagbag.py:_sync_to_db:710 > dag.py:bulk_write_to_db:3324 > manager.py:create_datasets:56:	2 -> new
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:618:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:893 > processor.py:update_import_errors:625:	1
	test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:689 > processor.py:_validate_task_pools:668 > pool.py:get_pools:69:	1
	python.py:pytest_pyfunc_call:162 > test_processor.py:test_counter_for_last_num_of_db_queries:1022 > test_processor.py:_process_file:114 > processor.py:process_file:904 > processor.py:update_dag_warnings:691:	1

```
As I see the mostly new queries come from `dagbag`. Maybe as a solution we can change the event handler from `session` to `engine`  and it will make our counter more accurate.

Reply:
Yeah. If we can use engine, then yeah, that would be a good one :).

I think you can do it independently from that PR?

Reply:
@potiuk yes I can do it in another PR

Reply:
The context manager approach I introduced looks like a very nice approach for that BTW. I thinl the way I changed it where each DB method counts queries separately (also to account for the fact it can be remote). And there actually - session is perfectly fine). The DAGBag parsing should be additionally wrapped with "engine" event I think to get the queries counted while parsing (but it should be done conditionally - only when database isolation is disabled (because there - there will be no engine to register event in the dag file processor. 

So you will likely need a context manager and wrap parsing with it and that context manager should register event and retrieve the count if db_isolation is off. 

Does it make sense @MaksYermak  ? Would you follow-up with it (I am now working on runinng db isolation tests and we will need that one rather quickly if we want to make it for Airflow 2.10.

Reply:
Perfect. I will merge it then and follow up with the rest of AIP-44.